### ISSUE TITLE
Historical information: Diminishing sample rate

### ISSUE THREAD

**Issue by koraa** [firehol/netdata#1613]
Hi,
it would be nice to store some historical data.  Of course, with long data retention times and a high sample rate storage requirements become ridiculous, but at a sampling rate of half an hour we could easily store 10y worth of data. That would not be very realtime though.

Thus I propose a feature that progressively decreases sample rates, the older data gets.

**Comment by koraa** [comment]
NOTE: I this could even be accomplished by using sparse db files and just zeroing data. Or cloning the same sample and compressing the file. (I am using transparent compression here, so in my case that would probably get 80% of that done).

**Comment by ktsaou** [comment]
I think historical data should not be stored on the servers. Saving or querying such a long history directly on the servers, will affect the applications running on them.

My goal is to properly have a few days, up to a week of data on each server, with minimal disk activity. To accomplish this, I plan to:

1. allow us to configure the maximum storage for metrics (i.e. 1GB of disk space)
2. store metrics on pages (i.e. 4KB each - adapted to match the machine page size)
3. deduplicate pages that are identical (there are lots of metrics that are just constant - gain 50%)
4. based on the statistics of each page, use either 2 or 4 bytes per metric (there are many metrics that can perfectly be stored in 2 bytes)
5. compress the pages before storing them on disk (gain 50%)
6. distribute the saving of pages in time, so that there will be no disk utilization spike

So, for each metric with full per-second detail, we will need to store 1KB every 1000 seconds.

In other words, for 1000 metrics we will fit **1 week** of data in 580MB of disk, and we will have 1KB/s constant disk writes on the server. For 2000 metrics, we will need to double everything.

This is too complex though. It will be done at some point, but I am not there yet...

For the moment, I suggest to use a back-end database (graphite, opentstb, prometheus) to centralize all the data.

**Comment by koraa** [comment]
Hi, I do agree that storing the data at full precision would be to harsh to query.

My thought here was to store that data in a way that it takes so little space it won't be a problem. Though you are probably right to question whether this should be even in the scope of this project.

Btw, did you consider actually applying any form of lossy compression to the time series data? Even after correcting for bits of the files that haven't been written, I am still getting compression ratios of >12x. With lossy compression I could imagine, that storing old data could become a non-issue.

I don't think any of the graph/time series dbs you mentioned provide such a feature, and storage space is really my main concern here.

**Comment by ktsaou** [comment]
> My thought here was to store that data in a way that it takes so little space it won't be a problem.

Yeah, I agree. However, I don't think there is any use for metrics with a detail less than 10 seconds. Everything above 10 seconds is business reporting, not performance monitoring... Actually, it is also very hard for me to work with `update every` set to 2 seconds. You cannot know that what you see is what it actually happens...

> Btw, did you consider actually applying any form of lossy compression to the time series data?

No, I have not. It depends on what the loss would be. If 899 is converted to 900 I think it would be ok, but if 899 is converted to 400 or 2000, then the loss is unacceptable. This is why lossy codecs work on a context that allows them to control the influence, the loss (i.e. music, images, etc).

I think the only acceptable lossy compression for netdata, is increasing `update_every`, i.e. merging multiple points to one by averaging them...

> I don't think any of the graph/time series dbs you mentioned provide such a feature, and storage space is really my main concern here.

They can average multiple points to one, and netdata can send them reduced too (so netdata can apply the lossy algorithm before sending them)...

**Comment by koraa** [comment]
I see, then I might add my specific use case here, to explain why I for me a lower resolution would be acceptable: My firm is evaluating switching to samba instead of smb; I am responsible for the server. We are right now doing load tests using long running background jobs on the server, progressively increasing the load over time. I am trying to find peak usage times over longer periods for time and I am trying to estimate how the load scales over time with the number of users. (Might not be linear, who knows). Since our jobs are always running for >30s I don't really care about a jitter of a few seconds

**Comment by ktsaou** [comment]
ok. keep in mind though that 1 min average (which is a lot lower to half an hour average), might not reveal the load your servers will experience. When you average things, you should always keep track of `min` and `max` too, since `max` might not be just twice the average. In such long durations, it might be 10 or 20 times the average!
